### ISSUE TITLE
About customizing ray directions

### ISSUE THREAD

**Issue by IwakuraRein** [tunabrain/tungsten#71]
Hi Benedit.

I know the path tracing algorithm demands the ωo generated randomly, I want to customize the ωo of the **first** bounce of each ray though. 

I've tried to fathom tungsten's core codes but can't get everything clear, for I just started learning pbrt. I've made PathTracer.cpp load an exr file describing the ωo of each pixel, but I have no idea where I can assign them. I guess the random generation happens in the class PathSampleGenerator, and handleSurface(), estimateDirect(), sampleDirect(), and chooseLight() in the TraceBase.cpp need modifications. Could you teach me a bit?

Thanks. 😊

**Comment by PhanSangTheAlerianLord** [comment]
I am surprise that no one even care to answer this question. My answer for your problem is you don't have to. It  is next to impossible to learn tungsten render or any big projects on github just by reading the code. Because you don't have a look into the process of making this project, so you don't know why function were written in that way. You will even struggle to understand the meaning of some variables. 
Tldr, If you try to learn tungsten render to write your own render. it's like someone throw you in the middle of Amazon and you have to find the way out. I myself have write a render that cover most of matt phar' s pbrt, but I still struggle to understand simpler code from other projects. A project without document is not worth studying.

If you want to learn ray tracing, this is the steps you should take:

1. Ray Tracing in One Weekend and finish the trio book. 
2. From that point, write your own 3d.obj parser.
3. Upgrade the bvh to sah bvh.
4. convert recursive path tracing to iterative path tracing
5. improve material to microfact model
6. adding diectional lighting, enviroment, fog..

And if you got there:

7. improve path tracing to bdpt, mtlt...
or
7. upgrade your project to cuda.

Have a good day!

**Comment by PhanSangTheAlerianLord** [comment]
No reply, great! It's suck to have such a detail answer and no reply. People these days forget about gratitude, I will quit answering online question from now on!

**Comment by PhanSangTheAlerianLord** [comment]
Fuck you all.
